![image](https://github.com/JasonLinjc/EPInformer/assets/43600523/7454f7fc-6fe1-4bf3-ab25-fd70a98d3e6c)<p align="center">
  <img height="500" src="images/EPInformer.png">
</p>

Welcome to the EPInformer framework repository! EPInformer is a scalable deep learning framework for gene expression prediction by integrating promoter-enhancer sequences with epigenomic signals. EPInformer excels in three key applications: 1) Accurately predicting gene expression levels using promoter-enhancer sequences, epigenomic signals, and chromatin contacts; 2) Efficiently identifying cell-type-specific enhancers, validated by CRISPR perturbation experiments; 3) Precisely predicting enhancer activity and identifying transcription factor binding motifs from sequences. These applications demonstrate EPInformer's versatility and effectiveness in gene expression regulation and enhancer biology analysis.

This repository can be used to run the EPInformer model and get the EPInformer predicted gene-enhancer interaction and gene expression predictions for input sequences and epigenomic signals

We also provide information and instructions for how to train the EPInformer deep learning sequence model.

### Requirements

EPInformer requires Python 3.6+ and Python packages PyTorch (>=2.0). You can follow PyTorch installation steps [here](https://pytorch.org/get-started/locally/).

### Setup
